### PR TITLE
OAK-10659 : remove orphaned nodes/documents

### PR DESCRIPTION
**Pull request by stefan-egli**

### CODE REVIEW

**Comment by stefan-egli** [comment]
Note that this PR also contains some new tests that were discovered while testing for orphaned node removal : lateWriteRemoveChildGC_noSweep (part of OAK-10535), lateWriteRemoveChildGC_withSweep (part of OAK-10658)

**Comment by stefan-egli** [comment]
```
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    BranchCommitGCTest.unmergedAddAndRemoveChild:372 expected:<0> but was:<1>
Error:  Errors: 
Error:    BranchCommitGCTest.unmergedAddChildren:100 NullPointer
Error:    BranchCommitGCTest.unmergedAddTwoChildren:228 NullPointer
[INFO] 
Error:  Tests run: 1893, Failures: 1, Errors: 2, Skipped: 106
```

looking into [that](https://github.com/apache/jackrabbit-oak/actions/runs/8005863809/job/21866285401?pr=1315) next

**Comment by stefan-egli** [comment]
fix attempt in https://github.com/apache/jackrabbit-oak/pull/1315/commits/21fc93adcf079ed31e8bae99c0ce4fccd91bb406

**Comment by stefan-egli** [comment]
PS: unmerged branch commit GC might need some rework/review after introduction of orphaned node deletion GC

**Comment by stefan-egli** [comment]
PS: now the flaky test succeeded but the build failed with the known
```
The plugin [java] does not support Java 11.0.22
```

I'd suggest to ignore this failure for now, merge and work on fixing the build in general in the branch

**Review by Joscorbe — APPROVED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 1]
```diff
@@ -942,21 +949,63 @@ public void collectGarbage(final NodeDocument doc, final GCPhases phases) {
...
+         * deletion cases that have not otherwise been deleted already (eg by DeletedDocsGC).
+         *
+         * @param doc
+         * @param traversedState
+         * @param phases
+         * @return true if the node is orphaned (and/or can be removed), false
+         *         otherwise
+         */
+        private boolean isDeletedOrOrphanedNode(NodeDocument doc, NodeState traversedState,
+                GCPhases phases) {
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]:
```suggestion
        private boolean isDeletedOrOrphanedNode(NodeState traversedState) {
```
Only traversedState is used.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 2]
```diff
@@ -942,21 +949,63 @@ public void collectGarbage(final NodeDocument doc, final GCPhases phases) {
...
-            collectOldRevisions(doc, phases, op);
-            // only add if there are changes for this doc
-            if (op.hasChanges()) {
+            // traversed state == state of node at doc.id based on head revision
+            NodeState traversedState = root;
+            for (String name : doc.getPath().elements()) {
+                traversedState = traversedState.getChildNode(name);
+            }
+
+            if (isDeletedOrOrphanedNode(doc, traversedState, phases)) {
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]:
```suggestion
            if (isDeletedOrOrphanedNode(traversedState)) {
```

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 3]
```diff
@@ -1255,28 +1305,40 @@ public void removeGarbage(final VersionGCStats stats) {
...
-                    stats.deletedPropsCount += deletedProps;
-                    stats.deletedUnmergedBCCount += deletedUnmergedBCSet.size();
-
-                    if (log.isDebugEnabled()) {
-                        log.debug("Updated [{}] documents, deleted [{}] properties, deleted [{}] unmergedBranchCommits",
-                                updatedDocs, deletedProps, deletedUnmergedBCSet.size());
+
+                    if (!orphanOrDeletedRemovalList.isEmpty()) {
+                        ds.remove(NODES, orphanOrDeletedRemovalList);
+                        final int removedSize = orphanOrDeletedRemovalList.size();
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: I would use the return value of `ds.remove` as removedSize not the actual list. 
There might be a case where db cannot delete all the passed documents.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 4]
```diff
@@ -1255,28 +1305,40 @@ public void removeGarbage(final VersionGCStats stats) {
...
+                        stats.updatedDetailedGCDocsCount += removedSize;
+                        stats.deletedDocGCCount += removedSize;
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1315]: We need to save these stats as well in RevisionGCStats.documentsDeleted` or we could add other stats in detailedGcStats for orphan node deletion.

Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1315]: Please add debug logging about the orphan nodes deletions.

**Review by rishabhdaim — CHANGES_REQUESTED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 1 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: +1 - it was a left-over of earlier work.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 3 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: good point - will adjust

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 4 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1315]: Ok, will revisit. I was a bit lost about all the stats we have...

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java [hunk 5]
```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
...
+                Arrays.asList("/a/b/c/d", "/a/b/c/d/e/f"), 3, "/d");
+    }
+
+    /**
+     * Tests whether DetailedGC can delete a large amount of randomly
+     * created orphans (that were added in a late-write manner)
+     */
+    @Test
+    public void lateWriteCreateManyChildrenGC() throws Exception {
+        List<String> nonOrphans = Arrays.asList("/a", "/b", "/c");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java]:
```suggestion
        List<String> nonOrphans = List.of("/a", "/b", "/c");
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java [hunk 6]
```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
...
+    }
+
+    /**
+     * Tests whether DetailedGC can delete a whole subtree "/a/b/c/d/**" that was
+     * added via late-writes.
+     */
+    @Test
+    public void lateWriteCreateChildTreeGC() throws Exception {
+        doLateWriteCreateChildrenGC(Arrays.asList("/a", "/a/b/c"),
+                Arrays.asList("/a/b/c/d", "/a/b/c/d/e/f"), 3, "/d");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java]:
```suggestion
        doLateWriteCreateChildrenGC(List.of("/a", "/a/b/c"),
                List.of("/a/b/c/d", "/a/b/c/d/e/f"), 3, "/d");
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java [hunk 7]
```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
...
+        assertFalse(getChildeNodeState(store1, nearestParent, true).hasChildNode(nearestChildPath.getName()));
+    }
+
+    /**
+     * Tests whether DetailedGC properly deletes a late-written addChild "/grand/parent/a"
+     */
+    @Test
+    public void lateWriteCreateChildGC() throws Exception {
+        doLateWriteCreateChildrenGC(Arrays.asList("/grand/parent"),
+                Arrays.asList("/grand/parent/a"), 1, "/d");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java]:
```suggestion
        doLateWriteCreateChildrenGC(List.of("/grand/parent"),
                List.of("/grand/parent/a"), 1, "/d");
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java [hunk 8]
```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
...
+                orphans, orphans.size() + commonOrphanParents.size(), "/d");
+    }
+
+    @Test
+    @Ignore(value = "OAK-10535 : fails currently as uncommitted revisions aren't yet removed")
+    public void lateWriteRemoveChildGC_noSweep() throws Exception {
+        assumeTrue(fixture.hasSinglePersistence());
+        enableDetailedGC(store1);
+        createNodes("/a/b/c/d");
+        lateWriteRemoveNodes(Arrays.asList("/a/b"), null);
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java]:
```suggestion
        lateWriteRemoveNodes(List.of("/a/b"), null);
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java [hunk 9]
```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
...
+     * at base revision r2-0-1,r1-0-2,
+     * branch: null, commit revision: re-0-1]
+     */
+    @Test
+    @Ignore(value = "OAK-10658 : fails currently as invalidation is missing (in classic GC) after late-write-then-sweep-then-GC")
+    public void lateWriteRemoveChildGC_withSweep() throws Exception {
+        assumeTrue(fixture.hasSinglePersistence());
+        enableDetailedGC(store1);
+        createNodes("/a/b/c/d");
+        lateWriteRemoveNodes(Arrays.asList("/a/b"), "/foo");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java]:
```suggestion
        lateWriteRemoveNodes(List.of("/a/b"), "/foo");
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java [hunk 10]
```diff
@@ -1182,6 +1201,362 @@ public void testDeletedPropsAndUnmergedBCWithCollision() throws Exception {
...
+        // but we need that to happen in prod code, not test
+//        store1.getNodeCache().invalidateAll();
+
+        // creating /a/b/c again, below late-write-removed /a/b
+        // triggered a ConflictException
+        createNodes("/a/b/c/d/e");
+        getChildeNodeState(store1, "/a/b/c/d/e", true);
+    }
+
+    private void enableDetailedGC(DocumentNodeStore dns) throws IllegalAccessException {
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java:1474]: This method could be moved to `DetailGCHelper` class or we could use the existing method `DetailGCHelper.enableDetailGC` there by passing `VersionGarbageCollector`

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 3 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: addressed in https://github.com/apache/jackrabbit-oak/pull/1315/commits/74ef58d446f770bdfc2723b0039409ca963c3c4d

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 4 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1315]: done in https://github.com/apache/jackrabbit-oak/pull/1315/commits/74ef58d446f770bdfc2723b0039409ca963c3c4d

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1315]: done also in https://github.com/apache/jackrabbit-oak/pull/1315/commits/74ef58d446f770bdfc2723b0039409ca963c3c4d

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java (hunk 10 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java:1474]: good point, done in https://github.com/apache/jackrabbit-oak/pull/1315/commits/0cca811570536841ac531d4d4ad7c00d2d20bce6

**Comment by stefan-egli** [comment]
@rishabhdaim , thx for the review, applied all comments, please revisit, thx

**Review by rishabhdaim — APPROVED** [review]

**Comment by stefan-egli** [comment]
ah, you right away merged it, ok :)

**Pull request merged**